File: src/main/scala/org/openapitools/client/api/QrsProduct.scala
package org.openapitools.client.api

import argonaut._
import argonaut.EncodeJson._
import argonaut.DecodeJson._

import org.http4s.{EntityDecoder, EntityEncoder}
import org.http4s.argonaut._
import org.joda.time.DateTime


import QrsProduct._

case class QrsProduct (
  name: Option[String],
label: Option[String],
description: Option[String],
effectiveDate: Option[String],
untilDate: Option[String],
registrationStatus: Option[String],
version: Option[String],
qrsType: Option[String],
links: Option[QrsProductLinks],
items: Option[List[QrsItem]])

object QrsProduct {
  import DateTimeCodecs._

  implicit val QrsProductCodecJson: CodecJson[QrsProduct] = CodecJson.derive[QrsProduct]
  implicit val QrsProductDecoder: EntityDecoder[QrsProduct] = jsonOf[QrsProduct]
  implicit val QrsProductEncoder: EntityEncoder[QrsProduct] = jsonEncoderOf[QrsProduct]
}

File: src/cdisc_library_api/specs/qrs_product.clj
(ns cdisc-library-api.specs.qrs-product
  (:require [clojure.spec.alpha :as s]
            [spec-tools.data-spec :as ds]
            [cdisc-library-api.specs.qrs-product-links :refer :all]
            [cdisc-library-api.specs.qrs-item :refer :all]
            )
  (:import (java.io File)))


(def qrs-product-data
  {
   (ds/opt :name) string?
   (ds/opt :label) string?
   (ds/opt :description) string?
   (ds/opt :effectiveDate) string?
   (ds/opt :untilDate) string?
   (ds/opt :registrationStatus) string?
   (ds/opt :version) string?
   (ds/opt :qrsType) string?
   (ds/opt :_links) qrs-product-links-spec
   (ds/opt :items) (s/coll-of qrs-item-spec)
   })

(def qrs-product-spec
  (ds/spec
    {:name ::qrs-product
     :spec qrs-product-data}))
